spring-integration-gcp/src/main/java/org/springframework/integration/gcp/storage/GcsSession.java

```diff
@@ -0,0 +1,196 @@
+/*
+ *  Copyright 2017 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.integration.gcp.storage;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.google.cloud.WriteChannel;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.BucketInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.integration.file.remote.session.Session;
+import org.springframework.util.Assert;
+import org.springframework.util.StreamUtils;
+
+/**
+ * @author João André Martins
+ */
+public class GcsSession implements Session<BlobInfo> {
+
+	private Storage gcs;
+
+	private static final Log LOGGER = LogFactory.getLog(GcsSession.class);
+
+	private static final String SEPARATOR = "/";
+
+	public GcsSession(Storage gcs) {
+		Assert.notNull(gcs, "The GCS client can't be null.");
+		this.gcs = gcs;
+	}
+
+	@Override
+	public boolean remove(String path) throws IOException {
+		String[] tokens = getBucketAndObjectFromPath(path);
+		Assert.state(tokens.length == 1 || tokens.length == 2,
+				"Path must be in the form of [bucket] or [bucket]/[blob name]");
+
+		return tokens.length == 1
+				? this.gcs.delete(tokens[0])
+				: this.gcs.delete(tokens[0], tokens[1]);
+	}
+
+	@Override
+	public BlobInfo[] list(String path) throws IOException {
+		// TODO(joaomartins): Only supports listing buckets, not folders.
+
+		Collection<BlobInfo> blobs = new ArrayList<>();
+
+		for (Blob blob : this.gcs.list(path).iterateAll()) {
+			blobs.add(blob);
+		}
+
+		return blobs.toArray(new BlobInfo[blobs.size()]);
+	}
+
+	@Override
+	public void read(String source, OutputStream outputStream) throws IOException {
+		String[] tokens = getBucketAndObjectFromPath(source);
```

Comment:
Is a check needed here for bucket-only paths? A few lines down you use use `tokens[1]` but that wouldn't be there if the path is just a bucket.

Reply:
The reason I didn't add the check is because the SI infrastructure never called `read()` on buckets in my tests.
But I guess it can't hurt to add one.